community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    /**
+     * Release a write lock taking with {@link #tryWriteLock()}.
+     */
+    public void unlockWrite()
+    {
+        long s, n;
+        do
+        {
+            s = getState();
+            if ( (s & CNT_MASK) == 0 )
+            {
+                throwUnmatchedUnlockWrite( s );
+            }
+            n = nextSeq( s ) - CNT_UNIT;
+        }
+        while ( !compareAndSetState( s, n ) );
+    }
+
+    private void throwUnmatchedUnlockWrite( long s )
+    {
+        throw new IllegalMonitorStateException( "Unmatched unlockWrite: " + describeState( s ) );
+    }
+
+    private long nextSeq( long s )
+    {
+        return (s & SEQ_IMSK) + (s + 1 & SEQ_MASK);
+    }
+
+    /**
+     * Grab an exclusive lock if one is immediately available. Exclusive locks will invalidate any overlapping
+     * optimistic read lock, and make write locks block. If any write locks are currently taken, then the attempt to
+     * grab an exclusive lock will fail.
+     * <p>
+     * Successfully grabbed exclusive locks must always be paired with a corresponding {@link #unlockExclusive()}.
+     *
+     * @return {@code true} if we successfully got an exclusive lock, {@code false} otherwise.
+     */
+    public boolean tryExclusiveLock()
+    {
+        long s = getState();
+        return ((s & CNT_MASK) == 0) & ((s & EXCL_MASK) == 0) && compareAndSetState( s, s + EXCL_MASK );
+    }
+
+    /**
+     * Unlock the currently held exclusive lock, and atomically and implicitly take an optimistic read lock, as
+     * represented by the returned stamp.
+     *
+     * @return A stamp that represents an optimistic read lock, in case you need it.
+     */
+    public long unlockExclusive()
+    {
+        long s = initiateExclusiveLockRelease();
+        long n = nextSeq( s ) - EXCL_MASK;
+        compareAndSetState( s, n );
+        return n;
+    }
+
+    /**
+     * Atomically unlock the currently held exclusive lock, and take a write lock.
+     */
+    public void unlockExclusiveAndTakeWriteLock()
+    {
+        long s = initiateExclusiveLockRelease();
+        long n = nextSeq( s ) - EXCL_MASK + CNT_UNIT;
+        compareAndSetState( s, n );
+    }
+
+    private long initiateExclusiveLockRelease()
+    {
+        long s = getState();
+        if ( (s & EXCL_MASK) != EXCL_MASK )
+        {
+            throwUnmatchedUnlockExclusive( s );
+        }
+        return s;
+    }
+
+    private void throwUnmatchedUnlockExclusive( long s )
+    {
+        throw new IllegalMonitorStateException( "Unmatched unlockExclusive: " + describeState( s ) );
+    }
+
+    @Override
+    public String toString()
+    {
+        long s = getState();
+        return describeState( s );
+    }
+
+    private String describeState( long s )
+    {
+        long excl = s >>> CNT_BITS + SEQ_BITS;
+        long cnt = (s & CNT_MASK) >> SEQ_BITS;
+        long seq = s & SEQ_MASK;
+        StringBuilder sb = new StringBuilder( "OptiLock[E:" ).append( excl );
```

Comment:
Would like to see describe state more descriptive, so it will not require to open source code to decrypt it.